penrose-web/src/shapes/ShapeDef.ts

```diff
@@ -0,0 +1,376 @@
+//#endregion
+
+//#region Resampling
+// TODO: The current resampling logic is a bit costly. Alternative: map over all varying paths and call sampleField/Property?
+
+/**
+ * Resample all shapes properties using their samplers.
+ * @param shapes Old shapes
+ * @ignore
+ */
+const sampleShapes = (shapes: Shape[]): Shape[] =>
+  shapes.map((shape: Shape) => sampleShape(shape, findDef(shape.shapeType)));
+
+/**
+ * Resample all properties of one shape.
+ * @param shape existing shape
+ * @param shapeDef shape definition
+ * @ignore
+ */
+const sampleShape = (shape: Shape, shapeDef: ShapeDef): Shape => ({
+  ...shape,
+  properties: mapValues(shape.properties, (_: Value<number>, prop: string) =>
+    sampleProperty(prop, shapeDef)
+  ),
+});
+
+/**
+ * Sample a property based on a shape definition.
+ * @param property property name
+ * @param shapeDef shape definition
+ */
+const sampleProperty = (
+  property: string,
+  shapeDef: ShapeDef
+): Value<number> => {
+  const propModels: IPropModel = shapeDef.properties;
+  const sampler = propModels[property];
+  if (sampler) return sampler[1]();
+  else
+    throw new Error(
+      `${property} is not a valid property to be sampled for shape ${shapeDef.shapeType}.`
+    );
+};
+
+/**
+ * Sample varying fields, which are assumed to be positional values. They are sampled with the canvas dimensions.
+ * @param state State that contains a list of varying paths
+ */
+const sampleFields = ({ varyingPaths }: State): number[] => {
+  const fieldPaths = varyingPaths.filter(
+    ({ tag }: Path) => tag === "AccessPath" || tag === "FieldPath"
+  );
+  return randFloats(fieldPaths.length, canvasXRange);
+};
+
+const samplePath = (path: Path, shapes: Shape[]): Value<number> => {
+  // HACK: for access and field paths, sample within the canvas width
+  if (path.tag === "AccessPath" || path.tag === "FieldPath") {
+    return constValue("FloatV", randFloat(...canvasXRange));
+  }
+  // for property path, use the sampler in shapedef
+  else {
+    const [{ contents: subName }, field, prop] = path.contents;
+    const { shapeType } = safe(
+      shapes.find(
+        (s: any) => s.properties.name.contents === `${subName}.${field}`
+      ),
+      `Cannot find shape ${subName}.${field}`
+    );
+    const shapeDef = findDef(shapeType);
+    const sampledProp: Value<number> = sampleProperty(prop, shapeDef);
+    return sampledProp;
+  }
+};
+
+export const resampleOne = (state: State): State => {
+  // resample all the uninitialized and varying values
+  const { varyingPaths, shapes, uninitializedPaths, params } = state;
+  const varyingValues: Value<number>[] = varyingPaths.map((p: Path) =>
+    samplePath(p, shapes)
+  );
+  const uninitValues: Value<VarAD>[] = uninitializedPaths.map((p: Path) =>
+    valueNumberToAutodiff(samplePath(p, shapes))
+  );
+
+  // update the translation with all uninitialized values (converted to `Done` values)
+  const uninitExprs: TagExpr<VarAD>[] = uninitValues.map((v) => val2Expr(v));
```

Comment:
Can remove parens on lambda

Reply:
Generally, it's encouraged to add type annotations, which involve parens. So I think ESLint is okay with this.